stdlib/public/runtime/Metadata.cpp

```diff
@@ -427,6 +428,16 @@ namespace {
   };
 } // end anonymous namespace
 
+namespace swift {
+  struct StaticAssertGenericMetadataCacheEntryValueOffset {
```

Comment:
Why have a struct for this?  `static_assert` is permitted at global scope.

Reply:
Yes, `static_assert` can be global but I figured that as the member `Value` is private, I needed to use `friend` to make it accessible to the `offsetof` and `friend` needs to specify some scope (class/struct). Alternatively I could make it public but it is probably less ideal.

Reply:
Hmm, `offsetof` is a macro, I don't think that it ties into the scope.  However, we _could_ also move the `static_assert` into the body of the declaration in `runtime/metadata.cpp`.

Reply:
As far as I can tell by playing with the compiler, `offsetof`, which is I think is translated into `_builtin_offsetof` ([doc](https://clang.llvm.org/docs/LanguageExtensions.html#builtin-offsetof)), appears to respect the c++ class access modifiers (`private`) and not to work _inside_ the declaration of the referred-to class itself (due to the `incomplete type` error, which presumably the compiler thinks the type hasn't been fully declared until the closing `};` of the class declaration). 

1. Since `Value` is a private member of `VariadicMetadataCacheEntryBase`, an `offsetof` in the derived class `GenericCacheEntry` cannot access it. So the `offsetof` cannot be inside the `GenericCacheEntry` declaration.
2. Because `VariadicMetadataCacheEntryBase` is still an incomplete type inside its declaration, the `offsetof` cannot be inside `VariadicMetadataCacheEntryBase`, either.
3. Thus, it needs to be outside of both of those classes. Then the question is whether we go with a global scope (which would require making the member `public`) or another struct/class (which would require using `friend`.) We have the latter approach here.

Reply:
Ah, yeah, I don't think that we should be making the members public for this.  Thanks for the explanation!  I think that this makes sense then.